### PR TITLE
[arrow2] Merge arrow2 and datafusion latest

### PR DESCRIPTION
**Pull request by Igosuki**
# Which issue does this PR close?

None

 # Rationale for this change
Updates the arrow2 branch with the latest of arrow2 and datafusion master

# What changes are included in this PR?
Extension traits to keep the API consistent and make future datafusion master merging easier.
Internally, some methods start using `Chunk`, I haven't done any perf tests.

# Are there any user-facing changes?
RecordBatch is `datafusion::record_batch::RecordBatch` instead of `datafusion::arrow::record_batch::RecordBatch`

# Issues 
Some tests fail, I haven't fixed everything.
Notably, checking schema validity between columns and schemas in record batch fails for decimal tests because arrow creates a Decimal(32, 32) column for Int128Array : https://github.com/jorgecarleitao/arrow2/blob/main/src/datatypes/mod.rs#L295

### CODE REVIEW

**Comment by alamb** [comment]
I changed the base to arrow2 branch rather than master.
Epic work @Igosuki  L

**Comment by Igosuki** [comment]
Woops 😆

Le ven. 28 janv. 2022 à 17:04, Andrew Lamb ***@***.***> a
écrit :

File: datafusion/src/datasource/file_format/csv.rs [hunk 1]
```diff
@@ -102,25 +103,18 @@ impl FileFormat for CsvFormat {
...
-            // schemas.push(schema.clone());
-            // records_to_read -= records_read;
-            // if records_to_read == 0 {
-            //     break;
-            // }
-            //
-            // FIXME: return recods_read from infer_schema
-            schemas.push(schema.clone());
-            records_to_read -= records_to_read;
+            if records_read == 0 {
```
Comment (alamb) [inline datafusion/src/datasource/file_format/csv.rs:113]: 👍

File: datafusion/src/field_util.rs [hunk 2]
```diff
@@ -109,3 +111,321 @@ pub fn struct_array_from(pairs: Vec<(Field, ArrayRef)>) -> StructArray {
     let values = pairs.iter().map(|v| v.1.clone()).collect();
     StructArray::from_data(DataType::Struct(fields), values, None)
 }
+
+/// Imitate arrow-rs Schema behavior by extending arrow2 Schema
+pub trait SchemaExt {
```
Comment (alamb) [inline datafusion/src/field_util.rs:116]: this is cool

**Review by alamb — APPROVED** [review]
I took a brief walk through this code @Igosuki  -- specifically the commits you added. I must say it is a very cool piece of work (especially with the compatibility traits, and `RecordBatch`)

👍

**Comment by Igosuki** [comment]
Although the extension traits help, I think the subtle differences between arrow-rs and arrow2 such as in Array creation will make merging always a bit painful, especially when merging unit tests... Not sure this way will be sustainable in the long run.

**Comment by houqp** [comment]
Really nice work on both the extension trait and recordbatch struct :+1:

With regards to unit tests merging, @Jimexist did some prior work to help reduce conflicts at: https://github.com/apache/arrow-datafusion/pull/1588. But I agree with you that merging is a lot of work. I think the goal is not to maintain the in the long run, but rather close the gaps in the arrow2 milestones as fast as we could, then make a final decision on whether we want to merge this into master or not as a community.

**Comment by houqp** [comment]
I believe we should not use squash merge to maintain long running branches, so I am going to manually push these changes into the arrow2 branch.

**Comment by houqp** [comment]
@Igosuki if you are planning on fixing some of the tests in the short term, then let me know so I won't be stepping on your toes. Otherwise, I will help with the tests this weekend.

**Comment by Igosuki** [comment]
@houqp Some fixes here https://github.com/Igosuki/arrow-datafusion/tree/arrow_testfix_20220129 remain a parquet projection issue, the decimal type issue I mentioned here and a test that fails because it's not spilling any memory when it's expected that it will.

**Pull request merged**